abi/abi.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package abi
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/ava-labs/avalanchego/utils/set"
+
+	"github.com/ava-labs/hypersdk/codec"
+)
+
+type ABI struct {
+	Actions []Action `serialize:"true" json:"actions"`
+	Types   []Type   `serialize:"true" json:"types"`
+}
+
+var _ codec.Typed = (*ABI)(nil)
+
+func (ABI) GetTypeID() uint8 {
+	return 0
+}
+
+// Field represents a field in the VM (Application Binary Interface).
+type Field struct {
+	// Name of the field, overridden by the json tag if present
+	Name string `serialize:"true" json:"name"`
+	// Type of the field, either a Go type or struct name (excluding package name)
+	Type string `serialize:"true" json:"type"`
+}
+
+// Action represents an action in the VM.
+type Action struct {
+	ID     uint8  `serialize:"true" json:"id"`
+	Action string `serialize:"true" json:"action"`
+}
+
+type Type struct {
+	Name   string  `serialize:"true" json:"name"`
+	Fields []Field `serialize:"true" json:"fields"`
+}
+
+func NewABI(actions []codec.Typed) (ABI, error) {
+	vmActions := make([]Action, 0)
+	vmTypes := make([]Type, 0)
+	typesSet := set.Set[string]{}
+
+	for _, action := range actions {
+		actionABI, typeABI, err := describeAction(action)
+		if err != nil {
+			return ABI{}, err
+		}
+		vmActions = append(vmActions, actionABI)
+		for _, t := range typeABI {
+			if !typesSet.Contains(t.Name) {
```

Comment:
How do we handle the case that there are two different types with the same name ie. if they're from two different packages?

Reply:
We can't, or we will have to include go package name into the JSON, which will cause troubles implementing it in other languages. I decided to drop support for same-name structs

Reply:
Happy to discuss